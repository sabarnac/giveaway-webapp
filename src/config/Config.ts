import Participant, { Avatar } from "../store/Participant";
import ConfigJson from "./config.json";
import Random, { Engine } from "random-js";

/**
 * Interface for JSON object representing avatar details.
 */
interface AvatarJson {
  url: string;
  altText: string;
}

/**
 * Interface for JSON object representing participant details.
 */
interface ParticipantJson {
  name: string;
  avatar?: AvatarJson;
}

/**
 * Enumeration of possible animation speeds.
 */
export const enum AnimationSpeed {
  HALF = 0.5,
  ONE = 1,
  ONE_POINT_FIVE = 1.5,
  TWO = 2,
  FIVE = 5,
  TEN = 10,
  TWENTY = 20,
}

/**
 * Class representing the basic configuration of the application.
 */
class Config {
  private _randomGenerator: Random;
  private _messages: string[];
  private _allParticipants: Participant[];
  private _speed: AnimationSpeed;
  private _participantsPerMatch: AnimationSpeed;

  private getRandomGenerator = (): Random => {
    const randomEngine: Engine = Random.engines.mt19937();
    return new Random(randomEngine);
  };

  private getMessage = (): string[] => ConfigJson.messages;

  private createParticipant = (user: string | ParticipantJson): Participant =>
    typeof user === "string"
      ? new Participant(user)
      : new Participant(
          (<ParticipantJson>user).name,
          this.createAvatar((<ParticipantJson>user).avatar),
        );

  private createAvatar = (avatar?: AvatarJson): Avatar | undefined =>
    avatar ? new Avatar(avatar.url, avatar.altText) : undefined;

  private getParticipants = (): Participant[] =>
    ConfigJson.users.map(this.createParticipant);

  private getSpeed = (): AnimationSpeed => AnimationSpeed.ONE;

  private getParticipantsPerMatch = (): number =>
    ConfigJson.participantsPerMatch;

  constructor() {
    this._randomGenerator = this.getRandomGenerator();
    this._messages = this.getMessage();
    this._allParticipants = this.getParticipants();
    this._speed = this.getSpeed();
    this._participantsPerMatch = this.getParticipantsPerMatch();
  }

  /**
   * Get the random generator.
   * @return {Random}
   */
  public get randomGenerator(): Random {
    return this._randomGenerator;
  }

  /**
   * Get the list of winner/loser messages.
   * @return {string[]}
   */
  public get messages(): string[] {
    return this._messages;
  }

  /**
   * Get the list of all participants in the giveaway.
   * @return {Participant[]}
   */
  public get allParticipants(): Participant[] {
    return this._allParticipants;
  }

  /**
   * Get the speed multiplier for animations.
   * @return {AnimationSpeed}
   */
  public get speed(): AnimationSpeed {
    return this._speed;
  }

  /**
   * Set the speed multiplier for animations.
   * @param {AnimationSpeed} value The new speed multiplier value.
   */
  public set speed(value: AnimationSpeed) {
    this._speed = value;
  }

  /**
   * Get the number of participants per match.
   * @return {AnimationSpeed}
   */
  public get participantsPerMatch(): AnimationSpeed {
    return this._participantsPerMatch;
  }
}

export default new Config();
