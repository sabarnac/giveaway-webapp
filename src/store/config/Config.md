Assume the configuration in `src/config/config.json` is set as:

```json
{
  "name": "Foobar Giveaway",
  "seed": 1234567890,
  "participantsPerMatch": 2,
  "messages": [
    "#winner #loser 1",
    "#winner #loser 2",
    "#winner #loser 3",
    "#winner #loser 4"
  ],
  "users": ["FOO", "BAR", "FAZ"]
}
```

Usage:

```tsx
import AnimationSpeed from "./store/config/AnimationSpeed";
import Participant from "./store/round/match/participant/Participant";

const name: string = Config.name; // "Foobar Giveaway
const messages: string[] = Config.messages; // ["#winner #loser 1", "#winner #loser 2", "#winner #loser 3", "#winner #loser 4"]
const message: string = Config.getRandomMessage("foobar", ["barfoo"]); // "foobar barfoo 3"
const allParticipants: Participant[] = Config.allParticipants; // [Participant("FOO"), Participant("BAR"), Participant("FAZ")]
const participantsPerMatch: number = Config.participantsPerMatch; // 2
const speed: number = Config.speed; // 1
Config.setSpeed(AnimationSpeed.TWO);
```
