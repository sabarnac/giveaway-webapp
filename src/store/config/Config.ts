import Participant from "../round/match/participant/Participant";
import ConfigJson from "./config.json";
import { observable, computed, action } from "mobx";
import RandomGenerator from "./RandomGenerator";
import AnimationSpeed from "./AnimationSpeed";

/**
 * Interface for JSON object representing avatar details.
 */
interface AvatarJson {
  /** The URL of the avatar. */
  url: string;
  /** The alternate text of the avatar image. */
  altText: string;
}

/**
 * Interface for JSON object representing participant details.
 */
interface ParticipantJson {
  /** The name of the participant. */
  name: string;
  /** The avatar of the participant (if present). */
  avatar?: AvatarJson;
  /** The weight of the participant (if present). */
  weight?: number;
}

/**
 * Class representing the basic configuration of the application.
 */
export default class Config {
  /** The name of the tournament. */
  private _name: string;
  /** A list of all messages. */
  private _messages: string[];
  /** The list of participants in the tournament. */
  private _allParticipants: Participant[];
  /** The number of participants per match. */
  private _participantsPerMatch: number;
  /** The animation speed multiplier. */
  @observable private _speed: number;
  /** The list of unused messages. */
  private _unusedMessages: string[];

  /** The singleton instance of the class, or null if not yet created. */
  private static _instance: Config | null = null;

  /**
   * Returns the name of the tournament.
   * @return {string} The tournament name.
   */
  private _getName = (): string => ConfigJson.name;

  /**
   * Returns the match conclusion messages that can be used for the tournament.
   * @return {string[]} The tournament match conclusion messages.
   */
  private _getMessages = (): string[] => ConfigJson.messages;

  /**
   * Returns a created participant for the given JSON details.
   * @param {string | ParticipantJson} user The participants details.
   * @return {Participant} The created participant.
   */
  private _createParticipant = (user: string | ParticipantJson): Participant =>
    typeof user === "string"
      ? new Participant(user)
      : new Participant(
          (<ParticipantJson>user).name,
          (<ParticipantJson>user).avatar,
          (<ParticipantJson>user).weight,
        );

  /**
   * Returns a created list of participants using the given participant details.
   * @return {Participant[]} The created participants list.
   */
  private _getParticipants = (): Participant[] =>
    ConfigJson.users.map(this._createParticipant);

  /**
   * Returns the max. number of participants that are to be present per match in the tournament.
   * @return {number} The max. number of participants per match.
   */
  private _getParticipantsPerMatch = (): number =>
    ConfigJson.participantsPerMatch;

  /**
   * Returns the animation speed multiplier of the application.
   * @return {number} The animation speed multiplier.
   */
  private _getSpeed = (): number => AnimationSpeed.get("ONE") as number;

  /**
   * Returns whether a string should be prepended to a participants' name.
   * @param {number} index The index of the participant.
   * @param {number} participantSize The total number of participants.
   * @return {boolean} Whether the string should be prepended.
   */
  private _shouldPrepend = (index: number, participantSize: number): boolean =>
    index === participantSize - 1 && participantSize > 1;

  /**
   * Returns a function that can returns the participant name prepended with a string if it's the last participant, or just the name.
   * @param {number} participantSize The total number of participants.
   * @param {string} stringToPrepend The string to prepend.
   * @return {(name: string, index: number) => string} The function that will return the participants' name
   */
  private _prependToLastParticipant = (
    participantSize: number,
    stringToPrepend: string,
  ): ((name: string, index: number) => string) => (
    name: string,
    index: number,
  ): string =>
    this._shouldPrepend(index, participantSize)
      ? `${stringToPrepend} ${name}`
      : name;

  /**
   * Returns a formatted losers list.
   * @param {string[]} names The list of loser names.
   * @return {string} The formatted string.
   */
  private _formatLosers = (names: string[]): string =>
    names.map(this._prependToLastParticipant(names.length, "and")).join(", ");

  private constructor() {
    this._name = this._getName();
    this._messages = this._getMessages();
    this._unusedMessages = RandomGenerator.shuffle([...this._messages]);
    this._allParticipants = this._getParticipants();
    this._participantsPerMatch = this._getParticipantsPerMatch();
    this._speed = this._getSpeed();
  }

  /**
   * Get an instance of the config, which is a singleton.
   * @return {Config} The instance of a config.
   */
  public static getInstance(): Config {
    return Config._instance
      ? Config._instance
      : (Config._instance = new Config());
  }

  /**
   * Get the name of the tournament.
   * @return {string} The tournament name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Get the list of winner/loser messages.
   * @return {string[]} The list of mesasges.
   */
  public get messages(): string[] {
    return [...this._messages];
  }

  /**
   * Returns the formatted version of the given message.
   * @param {string} message The message to format.
   * @param {string} winnerName The name of the winner.
   * @param {string[]} loserNames The names of the losers.
   * @return {string} The formatted message.
   */
  public getFormattedMessage = (
    message: string,
    winnerName: string,
    loserNames: string[],
  ): string =>
    message
      .replace("#winner", winnerName)
      .replace("#loser", this._formatLosers(loserNames));

  /**
   * Gets a random message with the winner and loser.
   * @return {string} The formatted message.
   */
  public getRandomMessage(): string {
    if (this._unusedMessages.length === 0) {
      this._unusedMessages = RandomGenerator.shuffle([...this._messages]);
    }
    const message = RandomGenerator.pick(this._unusedMessages);
    this._unusedMessages.splice(this._unusedMessages.indexOf(message), 1);
    return message;
  }

  /**
   * Gets the index of the given message.
   * @param {string} message The message.
   * @return {number} The index of the message.
   */
  public getMessageIndex = (message: string): number =>
    this._messages.indexOf(message);

  /**
   * Get the list of all participants in the tournament.
   * @return {Participant[]} The list of all participants.
   */
  public get allParticipants(): Participant[] {
    return this._allParticipants;
  }

  /**
   * Get the number of participants per match.
   * @return {number} The number of participants per match.
   */
  public get participantsPerMatch(): number {
    return this._participantsPerMatch;
  }

  /**
   * Get the speed multiplier for animations.
   * @return {number} The speed multiplier value.
   */
  @computed public get speed(): number {
    return this._speed;
  }

  /**
   * Set the speed multiplier for animations.
   * @param {number} value The new speed multiplier value.
   */
  @action public setSpeed(value: number): void {
    if (AnimationSpeed.hasValue(value)) {
      this._speed = value;
    } else {
      console.error(`Illegal value '${value}' for speed`);
    }
  }
}
