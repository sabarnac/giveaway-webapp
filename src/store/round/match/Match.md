Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const match: Match = new Match(
  [new Participant("FOO"), new Participant("BAR")],
  "1",
);
const matchId: string = match.id; // "1"
const matchCompleteId: string = match.fullId; // "1:1"
const participants: Participant[] = round.participants; // [Participant("FOO"), Participant("BAR")]
const losers: Participant[] = round.losers; // [Participant("BAR")]
const winner: Participant[] = round.winner; // Participant("FOO")
```
