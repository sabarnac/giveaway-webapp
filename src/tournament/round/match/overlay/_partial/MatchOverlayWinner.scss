.match-overlay__winner {
  padding: 4rem 4rem 2rem 4rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  opacity: 0.000001;
  &--entering,
  &--entered {
    opacity: 0.000001;
  }
  &--entering,
  &--entered {
    opacity: 0.999999;
  }
  & > .participant-entry {
    margin-bottom: 3rem;
  }
  & > h3,
  & > h5,
  & > h5 > strong,
  & > h5 > strong > em {
    text-align: center;
  }
}
