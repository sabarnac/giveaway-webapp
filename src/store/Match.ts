import { observable, computed, action } from "mobx";
import Participant from "./Participant";
import Random from "random-js";
import Config from "../config/Config";

/**
 * Class containing details of a match.
 */
export default class Match {
  @observable private _participants: Participant[];
  @observable private _winner: Participant;
  private static _RANDOM_GENERATOR: Random = Config.randomGenerator;

  private _getWinner = (): Participant =>
    Match._RANDOM_GENERATOR.pick(this._participants);

  public constructor(participants: Participant[]) {
    this._participants = participants;
    this._winner = this._getWinner();
  }

  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the list of losers of the match.
   * @returns {Participant[]} The list of losers.
   */
  @computed public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  public get winner(): Participant {
    return this._winner;
  }
}
