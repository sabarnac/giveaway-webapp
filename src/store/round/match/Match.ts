import Participant from "./participant/Participant";
import RandomGenerator from "../../config/RandomGenerator";
import Config from "../../config/Config";
const gcd = require("gcd");

/**
 * Class containing details of a match.
 */
export default class Match {
  /** The ID of the match. */
  private _id: string;
  /** The ID of the round the match is a part of. */
  private _roundId: string;
  /** The list of participants of the match. */
  private _participants: Participant[];
  /** The winner of the match. */
  private _winner: Participant;
  /** The match conclusion message. */
  private _message: string;
  /** The GCD of all participant weights. */
  private _participantWeightGcd: number;

  /** The config of the application. */
  private _config: Config;

  /** A counter for generating a unique ID for the match. */
  private static counter: { [k: string]: number } = {};

  /**
   * Returns the weight of the given participant.
   * @param {Participant} participant The participant to get the weight of.
   * @return {number} The participant weight.
   */
  private _getParticipantWeight = (participant: Participant): number =>
    participant.weight;

  /**
   * Returns the GCD of the weights of all participants of the match.
   * @return {number} The GCD of the participant weights.
   */
  private _getParticipantWeightsGcd = (): number =>
    this._participants
      .map(this._getParticipantWeight)
      .reduce(
        (prevGcd: number, weight: number): number => gcd(prevGcd, weight),
      );

  /**
   * Returns a list of cloned participants whose count is the weight of the participant divided by the GCD of weights of all participants in the list.
   * @param {Participant} participant The participant to clone.
   * @return {Participant[]} The list of cloned participants.
   */
  private _cloneParticipantByWeight = (
    participant: Participant,
  ): Participant[] => {
    const participants: Participant[] = [];
    for (let i = 0; i < participant.weight / this._participantWeightGcd; i++) {
      participants.push(participant);
    }
    return participants;
  };

  /**
   * Picks and returns a winner of the match from the list of participants at random, with participant weight factored in.
   * @return {Participant} The participant selected as the winner.
   */
  private _getWinner = (): Participant =>
    RandomGenerator.pick(
      this._participants.flatMap(this._cloneParticipantByWeight),
    );

  /**
   * Returns a random conclusion message for the match.
   * @return {string} The conclusion message.
   */
  private _getMessage = (): string => this._config.getRandomMessage();

  /**
   * Returns a unique ID for the match based on the round it belongs to.
   * @param {string} roundId The ID of the round the match belongs to.
   * @return {string} The unique match ID.
   */
  private _getMatchId = (roundId: string): string => {
    if (!Match.counter[roundId]) {
      Match.counter[roundId] = 1;
    }
    return `${Match.counter[roundId]++}`;
  };

  public constructor(
    config: Config,
    participants: Participant[],
    roundId: string,
  ) {
    this._config = config;

    this._id = `${this._getMatchId(roundId)}`;
    this._roundId = roundId;
    this._participants = participants;
    this._participantWeightGcd = this._getParticipantWeightsGcd();
    this._winner = this._getWinner();
    this._message = this._getMessage();
  }

  /**
   * Returns whether the participant is the winner of the match or not.
   * @param  {Participant} participant The participant to check.
   * @return {boolean} Whether the participant is the winner of the match or not.
   */
  private _isNotWinner = (participant: Participant): boolean =>
    !participant.equals(this._winner);

  /**
   * Get the match ID.
   * @return {string} The unique ID of the match within the context of the round it belongs to.
   */
  public get id(): string {
    return this._id;
  }

  /**
   * Get the complete match ID (prepended with round ID).
   * @return {string} The unique complete ID of the match.
   */
  public get fullId(): string {
    return `${this._roundId}:${this._id}`;
  }

  /**
   * Get the list of losers of the match.
   * @return {Participant[]} The list of losers.
   */
  public get losers(): Participant[] {
    return this._participants.filter(this._isNotWinner);
  }

  /**
   * Get the list of participants in the match.
   * @return {Participant[]} The list of participants.
   */
  public get participants(): Participant[] {
    return this._participants;
  }

  /**
   * Get the winner of the match.
   * @return {Participant} The winner.
   */
  public get winner(): Participant {
    return this._winner;
  }

  /**
   * Get the conslusion message of the match.
   * @return {string} The message.
   */
  public get message(): string {
    return this._message;
  }
}
