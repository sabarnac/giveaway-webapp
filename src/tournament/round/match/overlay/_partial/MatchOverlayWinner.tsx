import React, { useEffect } from "react";
import { Observer, inject } from "mobx-react";
import "./MatchOverlayWinner.scss";
import classNames from "classnames";
import ParticipantEntry from "../../participant/entry/ParticipantEntry";
import { CSSTransition } from "react-transition-group";
import {
  getNormalizedSpeed,
  runOnDelay,
  AnimationStateHook,
  useAnimationState,
  runOnPredicate,
  isInRange,
} from "../../../../../util/index";
import { RouteComponentProps, withRouter } from "react-router";
import Match from "../../../../../store/round/match/Match";
import {
  Trans,
  useTranslation,
  UseTranslationResponse,
  WithTranslation,
  withTranslation,
} from "react-i18next";
import Config from "../../../../../store/config/Config";
import Participant from "../../../../../store/round/match/participant/Participant";

/**
 * Properties of the match overlay winner React component.
 */
interface MatchOverlayWinnerProps {
  /** @ignore The application config. */
  config?: Config;
  /** CSS class */
  className: string;
  /** The current match details. */
  currentMatch: Match;
  /** Whether to show the winner or not. */
  show: boolean;
  /** Action to call when the view has finished showing the winner. */
  onWinnerComplete: () => void;
}

const STOP_QUERY_FLAG: string = "stop";

/**
 * React component for the match overlay winner.
 */
export default withTranslation()(
  inject("config")(
    withRouter(
      (
        props: MatchOverlayWinnerProps & RouteComponentProps & WithTranslation,
      ): JSX.Element => {
        const [
          currentState,
          updateState,
        ]: AnimationStateHook = useAnimationState();
        const shouldNotStopOnEnd: boolean = !new URLSearchParams(
          props.location.search,
        ).has(STOP_QUERY_FLAG);

        useEffect(runOnPredicate(currentState === 0, updateState));

        const { t }: UseTranslationResponse = useTranslation();
        const translatedMessagesList: string = t("matchOverlay.messages", {
          returnObjects: true,
          count: props.currentMatch.losers.length,
        });
        let matchMessage: string = props.currentMatch.message;

        const messageIndex: number = props.config!.getMessageIndex(
          props.currentMatch.message,
        );
        if (
          messageIndex !== -1 &&
          Array.isArray(translatedMessagesList) &&
          isInRange(messageIndex, 0, translatedMessagesList.length - 1)
        ) {
          matchMessage = translatedMessagesList[messageIndex];
        }

        const winnerName: string = props.currentMatch.winner.properName;
        const loserNames: string[] = props.currentMatch.losers.map(
          (loser: Participant): string => loser.properName,
        );

        return (
          <Observer>
            {() => (
              <CSSTransition
                in={currentState === 1 && props.show}
                timeout={getNormalizedSpeed(500)}
                classNames={{
                  enter: "",
                  enterActive: `${props.className}__winner--entering`,
                  enterDone: `${props.className}__winner--entered`,
                  exit: "",
                  exitActive: `${props.className}__winner--exiting`,
                  exitDone: `${props.className}__winner--exited`,
                }}
                mountOnEnter={true}
                unmountOnExit={true}
                onEntered={
                  shouldNotStopOnEnd
                    ? () =>
                        runOnDelay(
                          props.onWinnerComplete,
                          getNormalizedSpeed(4000),
                        )
                    : undefined
                }
              >
                <div
                  className={classNames(`${props.className}__winner`)}
                  style={{
                    transition: `opacity ${getNormalizedSpeed(
                      500,
                    )}ms ease-in-out`,
                  }}
                >
                  <Trans i18nKey="matchOverlay.winnerTitle">
                    <ParticipantEntry participant={props.currentMatch.winner} />
                    <h3>Won The Match!</h3>
                  </Trans>
                  <h5>
                    <strong>
                      <em>
                        {props.config!.getFormattedMessage(
                          matchMessage,
                          winnerName,
                          loserNames,
                        )}
                      </em>
                    </strong>
                  </h5>
                </div>
              </CSSTransition>
            )}
          </Observer>
        );
      },
    ),
  ),
);
