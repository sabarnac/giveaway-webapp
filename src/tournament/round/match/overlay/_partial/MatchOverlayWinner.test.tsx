import React from "react";
import MatchOverlayWinner from "./MatchOverlayWinner.react";
import { create, ReactTestRenderer } from "react-test-renderer";
import Match from "../../../../../store/round/match/Match";
import {
  createDummyMatch,
  createDummyConfig,
  createDummyTranslationProps,
  createDummyRouterProps,
} from "../../../../../util/test";
import Config from "../../../../../store/config/Config";

describe("MatchOverlayWinner Component.", () => {
  it("Matches snapshot.", () => {
    const config: Config = createDummyConfig();
    const match: Match = createDummyMatch();
    const component: ReactTestRenderer = create(
      <MatchOverlayWinner
        show={true}
        className="foobar"
        currentMatch={match}
        onWinnerComplete={() => {}}
        config={config}
        {...createDummyTranslationProps()}
        {...createDummyRouterProps()}
      />,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
