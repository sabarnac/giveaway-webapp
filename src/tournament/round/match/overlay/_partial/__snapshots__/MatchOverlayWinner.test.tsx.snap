// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchOverlayWinner Component. Matches snapshot. 1`] = `
<div
  classNames={
    Object {
      "enter": "",
      "enterActive": "foobar__winner--entering",
      "enterDone": "foobar__winner--entered",
      "exit": "",
      "exitActive": "foobar__winner--exiting",
      "exitDone": "foobar__winner--exited",
    }
  }
  in={false}
  mountOnEnter={true}
  onEntered={[Function]}
  timeout={500}
  unmountOnExit={true}
>
  <div
    className="foobar__winner"
    style={
      Object {
        "transition": "opacity 500ms ease-in-out",
      }
    }
  >
    <div
      className="participant participant-entry"
    >
      <div
        className="participant-entry__avatar"
      >
        <img
          alt="foobar-1"
          src="foobar-1"
        />
      </div>
      <div
        className="participant-entry__name"
      >
        <strong>
          Fozbaz 1
        </strong>
      </div>
    </div>
    <h3>
      Won The Match!
    </h3>
    <h5>
      <strong>
        <em>
          formatted foobar message
        </em>
      </strong>
    </h5>
  </div>
</div>
`;
